### PR TITLE
fs: optimize dvc list --recursive

### PR DESCRIPTION
**Pull request by rlamy**
See #8135 for context.

`._get_key()` is quite expensive. The general idea is to avoid flip-flopping between path and "key" representations.

### CODE REVIEW

File: dvc/fs/dvc.py [hunk 1]
```diff
@@ -137,12 +138,21 @@ def _getcwd():
         if hasattr(repo, "dvc_dir"):
             self._datafss[key] = DataFileSystem(index=repo.index.data["repo"])
 
-    def _get_key(self, path):
+    def _get_key(self, path) -> Key:
```
Comment (skshetry) [inline dvc/fs/dvc.py:141]: Would lru help here?

Comment (rlamy) [reply dvc/fs/dvc.py:141]: The goal is to perform this operation only once per path, so I guess not, but I'll keep it in mind.

**Comment by efiop** [comment]
For the record: discussed during team meetings that we can go step-by-step here and it is worth merging this considering the results that it brings already 🙂

**Review by efiop — APPROVED** [review]
🔥

**Pull request merged**